samples/apps/tinyRA/tinyra/tui.py

```diff
@@ -0,0 +1,1186 @@
+
+    def compose(self) -> ComposeResult:
+        yield DirectoryTree(self.dirpath)
+
+    def on_mount(self) -> None:
+        self.set_interval(1, self.update_dir_contents)
+
+    def update_dir_contents(self) -> None:
+        self.dir_contents = str(os.listdir(self.dirpath))
+
+    def watch_dir_contents(self):
+        self.query_one(DirectoryTree).reload()
+
+    def on_tree_node_highlighted(self, event: DirectoryTree.NodeHighlighted) -> None:
+        logging.info(f"Highlighted {event.node}")
+        self.highlighted_node = event.node
+
+
+class ChatDisplay(ScrollableContainer):
+    """
+    A container for displaying the chat history.
+
+    When a new message is detected, it is mounted to the container.
+    """
+
+    limit_history = 100
+
+    # num_messages = reactive(len(fetch_chat_history()))
+
+    # async def watch_num_messages(self) -> None:
+    # self.scroll_end()
+
+    def compose(self) -> ComposeResult:
+        chat_history = fetch_chat_history()
+        for message in chat_history[-self.limit_history :]:
+            widget = message_display_handler(message)
+            yield widget
+
+
+class ChatInput(Static):
+    """
+    A widget for user input.
+    """
+
+    def on_mount(self) -> None:
+        input = self.query_one(Input)
+        input.focus()
+
+    def compose(self) -> ComposeResult:
+        yield Input(id="chat-input-box")
+
+
+class QuitScreen(ModalScreen):
+    """Screen with a dialog to quit."""
+
+    BINDINGS = [("escape", "app.pop_screen", "Pop screen")]
+
+    def compose(self) -> ComposeResult:
+        yield Grid(
+            Static("Are you sure you want to quit?", id="question"),
+            Grid(
+                Button("Quit", variant="error", id="quit"),
+                Button("Cancel", variant="primary", id="cancel"),
+                id="quit-screen-footer",
+            ),
+            id="quit-screen-grid",
+        )
+
+    def on_button_pressed(self, event: Button.Pressed) -> None:
+        if event.button.id == "quit":
+            self.app.exit()
+        else:
+            self.app.pop_screen()
+
+
+class SettingsScreen(ModalScreen):
+    """Screen with a dialog to display settings."""
+
+    BINDINGS = [("escape", "app.pop_screen", "Dismiss")]
+
+    def compose(self) -> ComposeResult:
+        self.widget_user_name = Input(APP_CONFIG.get_user_name())
+        self.widget_user_bio = TextArea(APP_CONFIG.get_user_bio(), id="user-bio")
+        self.widget_user_preferences = TextArea(APP_CONFIG.get_user_preferences(), id="user-preferences")
+
+        tools = APP_CONFIG.get_tools()
+
+        with TabbedContent("User", "Tools", "History", id="settings-screen"):
+            # Tab for user settings
+            yield Container(
+                ScrollableContainer(
+                    Container(Label("User", classes="form-label"), self.widget_user_name),
+                    Container(Label("Bio", classes="form-label"), self.widget_user_bio),
```

Comment:
It was helpful for me have the perspective that both bio and preferences are part of instructions in the system message for assistant.  And that the assitant is only instructed to write code that it thinks I will execute.